amcl/test/basic_localization.py

```diff
@@ -21,61 +22,101 @@ def setUp(self):
+        self.tfBuffer = None
 
-    def tf_cb(self, msg):
-        for t in msg.transforms:
-            if t.header.frame_id == 'map':
-                self.tf = t.transform
-                (a_curr, a_diff) = self.compute_angle_diff()
-                print('Curr:\t %16.6f %16.6f %16.6f' % (self.tf.translation.x, self.tf.translation.y, a_curr))
-                print('Target:\t %16.6f %16.6f %16.6f' % (self.target_x, self.target_y, self.target_a))
-                print('Diff:\t %16.6f %16.6f %16.6f' % (
-                    abs(self.tf.translation.x - self.target_x), abs(self.tf.translation.y - self.target_y), a_diff))
-
-    def compute_angle_diff(self):
+    def print_pose_diff(self):
+        a_curr = self.compute_angle()
+        a_diff = a_curr - self.target_a
+        print('Curr:\t %16.6f %16.6f %16.6f' % (self.tf.translation.x, self.tf.translation.y, a_curr))
+        print('Target:\t %16.6f %16.6f %16.6f' % (self.target_x, self.target_y, self.target_a))
+        print('Diff:\t %16.6f %16.6f %16.6f' % (
+            abs(self.tf.translation.x - self.target_x), abs(self.tf.translation.y - self.target_y), a_diff))
+
+    def get_pose(self):
+        try:
+            tf_stamped = self.tfBuffer.lookup_transform("map", "base_link", rospy.Time())
+            self.tf = tf_stamped.transform
+            self.print_pose_diff()
+        except (tf2.LookupException, tf2.ExtrapolationException):
+            pass
+
+    @staticmethod
+    def wrap_angle_a(angle):
+        """
+        Wrap angle to [-pi, pi)
+        """
+        angle += pi
+        while angle < 0:
+            angle += 2*pi
+        return fmod(angle, 2*pi) - pi
+
+    @staticmethod
+    def wrap_angle_b(angle):
+        """
+        Wrap angle to [0, 2*pi)
+        """
+        while angle < 0:
+            angle += 2*pi
+        return fmod(angle, 2*pi)
+
+    def wrap_angle(self, angle):
+        """
+        Wraps angles according the target angle
+        If the target angle is in the righthand plane, wrap to [-pi, pi)
+        if the target angle is in the lefthand plane, wrap to [0, 2*pi)
+
+        :param angle: Angle to be wrapped
+        :return: wrapped angle
+        """
+        if abs(self.target_a) <= 0.5*pi or abs(self.target_a) >= 1.5*pi:
+            return self.wrap_angle_a(angle)
+        else:
+            return self.wrap_angle_b(angle)
+
+    def compute_angle(self):
         rot = self.tf.rotation
         a = PyKDL.Rotation.Quaternion(rot.x, rot.y, rot.z, rot.w).GetRPY()[2]
-        d_a = self.target_a
-
-        return (a, abs(fmod(a - d_a + 5*pi, 2*pi) - pi))
+        return self.wrap_angle(a)
 
     def test_basic_localization(self):
         global_localization = int(sys.argv[1])
         self.target_x = float(sys.argv[2])
         self.target_y = float(sys.argv[3])
-        self.target_a = float(sys.argv[4])
+        self.target_a = self.wrap_angle_a(float(sys.argv[4]))
+        self.target_a = self.wrap_angle(self.target_a)
```

Comment:
Why wrap twice? I feel like a single wrapping would be enough

Reply:
It would be this or adding extra logic in the `wrap_angle` function.